source/common/buffer/watermark_buffer.h

```diff
@@ -72,15 +74,98 @@ class WatermarkBuffer : public OwnedImpl {
 
 using WatermarkBufferPtr = std::unique_ptr<WatermarkBuffer>;
 
+class WatermarkBufferFactory;
+
+/**
+ * A BufferMemoryAccountImpl tracks allocated bytes across associated buffers and
+ * slices that originate from those buffers, or are untagged and pass through an
+ * associated buffer.
+ *
+ * This BufferMemoryAccount is produced by the *WatermarkBufferFactory*.
+ */
+class BufferMemoryAccountImpl : public BufferMemoryAccount {
+public:
+  // Used to create the account, and complete wiring with the factory
+  // and shared_this_.
+  static BufferMemoryAccountSharedPtr createAccount(WatermarkBufferFactory* factory,
+                                                    Http::StreamResetHandler* reset_handler);
+  ~BufferMemoryAccountImpl() override { ASSERT(buffer_memory_allocated_ == 0); }
```

Comment:
`ASSERT(reset_handler_ == nullptr);` despite it being impossible for this to not be the case given use of `shared_this_`.